distributed/tests/test_spill.py

```diff
@@ -11,72 +12,132 @@ def test_spillbuffer(tmpdir):
+    assert buf.slow.weight_by_key == {"d": pickle_size}
+    assert buf.slow.total_weight == pickle_size
 
     # Single key is larger than target and goes directly into slow
     e = "e" * 500
-    slarge = sizeof(e)
+    pickle_large_size = sum(len(frame) for frame in serialize_bytelist(e))
+
     buf["e"] = e
     assert set(buf.disk) == {"d", "e"}
-    assert buf.spilled_by_key == {"d": s, "e": slarge}
-    assert buf.spilled_total == s + slarge
+    assert buf.slow.weight_by_key == {"d": pickle_size, "e": pickle_large_size}
+    assert buf.slow.total_weight == pickle_size + pickle_large_size
 
     # Updating a spilled key with another larger than target updates slow directly
     buf["d"] = "d" * 500
     assert set(buf.disk) == {"d", "e"}
-    assert buf.spilled_by_key == {"d": slarge, "e": slarge}
-    assert buf.spilled_total == slarge * 2
+    assert buf.slow.weight_by_key == {"d": pickle_large_size, "e": pickle_large_size}
+    assert buf.slow.total_weight == pickle_large_size * 2
+
+
+def test_spillbuffer_maxlim(tmpdir):
+    buf = SpillBuffer(str(tmpdir), target=200, max_spill=600)
+    # Convenience aliases
+    assert buf.memory is buf.fast
+    assert buf.disk is buf.slow
+
+    assert not buf.slow.weight_by_key
+    assert buf.slow.total_weight == 0
+
+    a, b, c, d = "a" * 200, "b" * 100, "c" * 100, "d" * 200
+
+    psize_a = sum(len(frame) for frame in serialize_bytelist(a))  # this is size in disk
+
+    # size of a is bigger than target and is smaller than max_spill, key should be in slow
+    buf["a"] = a
+    assert not buf.memory
+    assert not buf.fast.weights
+    assert set(buf.disk) == {"a"}
+    assert buf.slow.weight_by_key == {"a": psize_a}
+    assert buf.slow.total_weight == psize_a
+    assert buf["a"] == a
+
+    # size of b is smaller than target key should be in fast
+    sb = sizeof(b)
+    buf["b"] = b
+    assert set(buf.memory) == {"b"}
+    assert buf.fast.weights == {"b": sb}
+    assert buf["b"] == b
+    assert buf.fast.total_weight == sb
+
+    # size of c is smaller than target but b+c > target, c should stay in fast and b move to slow
+    # since the max_spill limit has not been reached yet
+    sc = sizeof(c)
+    psize_b = sum(len(frame) for frame in serialize_bytelist(b))
+
+    buf["c"] = c
+    assert set(buf.memory) == {"c"}
+    assert buf.fast.weights == {"c": sc}
+    assert buf["c"] == c
+    assert buf.fast.total_weight == sc
+
+    assert set(buf.disk) == {"a", "b"}
+    assert buf.slow.weight_by_key == {"a": psize_a, "b": psize_b}
+    assert buf.slow.total_weight == psize_a + psize_b
+
+    # size of d > target, d should go to slow but slow reached the max_spill limit then d
+    # will end up on fast with c (which can't be move to slow because it won't fit either)
+    sd = sizeof(d)
+
+    buf["d"] = d
+    assert set(buf.memory) == {"c", "d"}
+    assert buf.fast.weights == {"c": sc, "d": sd}
+    assert buf["d"] == d
+    assert buf.fast.total_weight == sc + sd
```

Comment:
Please add test:
- overwrite a key that was in slow, but the size of the new key is larger than max_spill
- overwrite a key that was in fast, but the size of the new key is larger than max_spill